File: app/src/modules/ai/domain/docConversations.ts
// 文档级会话数据模型：只包含纯数据结构，不涉及 IO

export type SummaryLevel = 0 | 1 | 2

/** 文档会话类型：按 Markdown / PDF 区分，同一目录下互不干扰 */
export type DocConversationKind = 'markdown' | 'pdf'

export type DocConversationMessageMeta = {
  providerType?: 'dify' | 'openai' | 'local' | 'coze' | 'other'
  modelName?: string
  hasImage?: boolean
  tokensUsed?: number
  /** 0 表示原始消息，1/2 表示不同层级的摘要 */
  summaryLevel?: SummaryLevel
  /** 仅摘要消息使用：覆盖的原始/摘要消息 ID 列表 */
  coversMessageIds?: string[]
  /** 仅摘要消息使用：覆盖的时间范围 */
  coveredTimeRange?: { from: number; to: number }
}

// 单条消息（用于文档会话时间线与非 Dify 模型上下文来源）
export type DocConversationMessage = {
  id: string
  docPath: string
  timestamp: number
  role: 'user' | 'assistant' | 'system'
  content: string
  meta?: DocConversationMessageMeta
}

// 按 docPath 聚合的一份文档会话记录
// - kind 缺省时按 'markdown' 处理，用于兼容旧数据
export type DocConversationRecord = {
  docPath: string
  /** 会话类型：Markdown / PDF。旧记录缺省视为 markdown */
  kind?: DocConversationKind
  sessionId: string
  lastActiveAt: number
  /** 旧字段：仅保留用于兼容，未来将迁移至 difyProviderConversations */
  difyConversationId?: string
  /** 按 ProviderId 隔离的 Dify 会话 ID 映射 */
  difyProviderConversations?: Record<string, string>
  messages: DocConversationMessage[]
}

// 轻量索引结构：用于列表/菜单展示
export type ConversationIndexEntry = {
  docPath: string
  /** 会话类型：Markdown / PDF。旧记录缺省视为 markdown */
  kind?: DocConversationKind
  sessionId: string
  lastActiveAt: number
  /** 只要存在任一 Dify 会话 ID 即为 true */
  hasDifyConversation: boolean
  messageCount: number
}

File: app/src-tauri/src/doc_conversations.rs
use crate::{err_payload, new_trace_id, ok, ErrorCode, ResultPayload};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::PathBuf;
use tauri::{AppHandle, Manager};
use tokio::fs;

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct DocConversationMessageMetaCfg {
    #[serde(default)]
    pub provider_type: Option<String>,
    #[serde(default)]
    pub model_name: Option<String>,
    #[serde(default)]
    pub has_image: Option<bool>,
    #[serde(default)]
    pub tokens_used: Option<u32>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct DocConversationMessageCfg {
    pub id: String,
    pub doc_path: String,
    pub timestamp: i64,
    pub role: String,
    pub content: String,
    #[serde(default)]
    pub meta: Option<DocConversationMessageMetaCfg>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct DocConversationRecordCfg {
    pub doc_path: String,
    pub session_id: String,
    pub last_active_at: i64,
    #[serde(default)]
    pub dify_conversation_id: Option<String>,
    #[serde(default)]
    pub dify_provider_conversations: Option<HashMap<String, String>>,
    #[serde(default)]
    pub messages: Vec<DocConversationMessageCfg>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct ConversationIndexEntryCfg {
    pub doc_path: String,
    pub session_id: String,
    pub last_active_at: i64,
    pub has_dify_conversation: bool,
    pub message_count: usize,
}

fn ai_conversations_dir(app: &AppHandle) -> std::io::Result<PathBuf> {
    if let Ok(mut dir) = app.path().config_dir() {
        dir.push("haomd");
        dir.push("ai-conversations");
        std::fs::create_dir_all(&dir)?;
        Ok(dir)
    } else {
        let mut dir = std::env::current_dir()?;
        dir.push("ai-conversations");
        std::fs::create_dir_all(&dir)?;
        Ok(dir)
    }
}

fn ai_conversations_data_path(app: &AppHandle) -> std::io::Result<PathBuf> {
    let mut dir = ai_conversations_dir(app)?;
    dir.push("conversations_data.json");
    Ok(dir)
}

fn ai_conversations_index_path(app: &AppHandle) -> std::io::Result<PathBuf> {
    let mut dir = ai_conversations_dir(app)?;
    dir.push("conversations_index.json");
    Ok(dir)
}

async fn read_doc_conversations(app: &AppHandle) -> std::io::Result<Vec<DocConversationRecordCfg>> {
    let path = ai_conversations_data_path(app)?;
    match fs::read(&path).await {
        Ok(bytes) => {
            let records: Vec<DocConversationRecordCfg> =
                serde_json::from_slice(&bytes).unwrap_or_default();
            Ok(records)
        }
        Err(err) if err.kind() == std::io::ErrorKind::NotFound => Ok(vec![]),
        Err(err) => Err(err),
    }
}

async fn write_doc_conversations(
    app: &AppHandle,
    records: &[DocConversationRecordCfg],
) -> std::io::Result<()> {
    let data_path = ai_conversations_data_path(app)?;
    let data_bytes = serde_json::to_vec_pretty(records)?;
    fs::write(&data_path, data_bytes).await?;

    let index_entries: Vec<ConversationIndexEntryCfg> = records
        .iter()
        .map(|rec| ConversationIndexEntryCfg {
            doc_path: rec.doc_path.clone(),
            session_id: rec.session_id.clone(),
            last_active_at: rec.last_active_at,
            has_dify_conversation: rec
                .dify_conversation_id
                .as_ref()
                .map(|s| !s.is_empty())
                .unwrap_or(false)
                || rec
                    .dify_provider_conversations
                    .as_ref()
                    .map(|m| !m.is_empty())
                    .unwrap_or(false),
            message_count: rec.messages.len(),
        })
        .collect();

    let index_path = ai_conversations_index_path(app)?;
    let index_bytes = serde_json::to_vec_pretty(&index_entries)?;
    fs::write(&index_path, index_bytes).await?;

    Ok(())
}

#[tauri::command]
pub async fn load_doc_conversations(
    app: AppHandle,
) -> ResultPayload<Vec<DocConversationRecordCfg>> {
    let trace = new_trace_id();
    match read_doc_conversations(&app).await {
        Ok(records) => ok(records, trace),
        Err(err) => err_payload(
            ErrorCode::IoError,
            format!("读取 conversations_data 失败: {err}"),
            trace,
        ),
    }
}

#[tauri::command]
pub async fn save_doc_conversations(
    app: AppHandle,
    records: Vec<DocConversationRecordCfg>,
) -> ResultPayload<()> {
    let trace = new_trace_id();
    match write_doc_conversations(&app, &records).await {
        Ok(()) => ok((), trace),
        Err(err) => err_payload(
            ErrorCode::IoError,
            format!("写入 conversations_data 失败: {err}"),
            trace,
        ),
    }
}
